config/config.go

```diff
@@ -198,6 +203,130 @@ type HTTPRewriteResponses struct {
+		Province:           c.Province,
+		StreetAddress:      c.StreetAddress,
+		PostalCode:         c.PostalCode,
+	}
+}
+
+// VerifyClientCertMetadata checks for th certificate chain received from the tls.Config.VerifyPeerCertificate
+// function callback, whether any configured VerifyClientCertificateMetadataRule applies.
+//
+// If a rule does apply, it is evaluated.
+//
+// Returns an error if there is an applicable rule which does not find a valid client certificate subject.
+func VerifyClientCertMetadata(rules []VerifyClientCertificateMetadataRule, chains [][]*x509.Certificate, logger logger.Logger) error {
+	for _, rule := range rules {
+		for _, chain := range chains {
+			requiredSubject := rule.CASubject.ToName()
+			if !checkIfRuleAppliesToChain(chain, logger, requiredSubject) {
+				continue
+			}
+			err := checkClientCertificateMetadataRule(chain, logger, rule)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// checkIfRuleAppliesToChain checks, whether the provided certificate chain contains a CA certificate
+// whose subject matches the requiredCASubject name.
+//
+// Returns true, if a CA certificate in the chain (cert.IsCA == true) matches the requiredCASubject name.
+func checkIfRuleAppliesToChain(chain []*x509.Certificate, logger logger.Logger, requiredCASubject pkix.Name) bool {
+	for i, cert := range chain {
+		logger.Debug("cert", zap.Int("index", i), zap.Bool("ca", cert.IsCA), zap.String("subject", cert.Subject.String()), zap.String("issuer", cert.Issuer.String()))
+		if cert.IsCA && requiredCASubject.ToRDNSequence().String() == cert.Subject.ToRDNSequence().String() {
+			return true
+		}
+	}
+	return false
+}
+
+// checkClientCertificateMetadataRule is called by checkIfRuleAppliesToChain. When the CA subject matches, the subject
+// of the client certificate is compared agains the subjects defined in rule.
+//
+// Returns an error when:
+// * the certificate does not match any of the ValidSubjects in rule.
+// * the chain does not contain a client certificate (i.e. IsCA == false).
+func checkClientCertificateMetadataRule(chain []*x509.Certificate, logger logger.Logger, rule VerifyClientCertificateMetadataRule) error {
+	for _, cert := range chain {
+		if !cert.IsCA {
+			subject := cert.Subject
+			for _, validSubject := range rule.ValidSubjects {
+				vaildCertSubject := validSubject.ToName()
+				if vaildCertSubject.ToRDNSequence().String() == subject.ToRDNSequence().String() {
+					logger.Debug("chain", zap.String("issuer", cert.Issuer.String()), zap.Bool("CA", cert.IsCA), zap.String("subject", cert.Subject.String()))
+					return nil
+				}
+			}
+			logger.Warn("invalid-subject", zap.String("issuer", cert.Issuer.String()), zap.String("subject", cert.Subject.String()), zap.Object("allowed", rule.ValidSubjects))
+			return fmt.Errorf("subject not in the list of allowed subjects for CA Subject %q: %q", rule.CASubject, subject)
+		}
+	}
+	return fmt.Errorf("incomplete chain")
```

Comment:
1. Is there a test for this case?
2. Can you provide a more verbose error message? I don't think I would understand this message without some code digging to find your helpful comments.

Reply:
No there is no test yet.

Yes we will add a test and a more verbose error message. It is possible/likely that this will not get through the regular client certificate check already. In that case I would remove it.

Reply:
I have provided a more verbose error message and a comment.

I couldn't find a way to get to that point of having an incomplete chain processed by the callback for verifying the peer cert.